File: scripts/fix-phpstan-iterables.py
#!/usr/bin/env python3
"""Inject PHPStan level-8 array value types into PHPDoc blocks."""

from __future__ import annotations

import re
import sys
from pathlib import Path

ARRAY_TYPE = "array<string, mixed>"
ROOT = Path(__file__).resolve().parents[1]
TARGETS = [ROOT / "backend/app", ROOT / "backend/bootstrap"]


def has_array_type_tag(doc: str, kind: str, name: str | None = None) -> bool:
    if kind == "return":
        return bool(re.search(r"@return\s+array<", doc) or re.search(rf"@return\s+{re.escape(ARRAY_TYPE)}", doc))
    if kind == "param" and name:
        return bool(
            re.search(rf"@param\s+array<[^>]+>\s+\${re.escape(name)}\b", doc)
            or re.search(rf"@param\s+{re.escape(ARRAY_TYPE)}\s+\${re.escape(name)}\b", doc)
        )
    if kind == "var":
        return bool(re.search(r"@var\s+array<", doc) or re.search(rf"@var\s+{re.escape(ARRAY_TYPE)}", doc))
    return False


def inject_tags(doc: str, params: list[str], returns_array: bool) -> str:
    doc = doc.rstrip()
    if doc.endswith("*/"):
        body = doc[:-2].rstrip()
    else:
        body = doc

    for param in params:
        if has_array_type_tag(body, "param", param):
            continue
        body += f"\n * @param {ARRAY_TYPE} ${param}"

    if returns_array and not has_array_type_tag(body, "return"):
        body += f"\n * @return {ARRAY_TYPE}"

    return body + "\n */"


def patch_properties(content: str) -> str:
    def repl(match: re.Match[str]) -> str:
        indent, prop = match.group(1), match.group(2)
        before = content[: match.start()]
        window = before[-400:]
        if re.search(r"@var\s+array<", window) or re.search(rf"@var\s+{re.escape(ARRAY_TYPE)}", window):
            return match.group(0)
        return f"{indent}/** @var {ARRAY_TYPE} */\n{indent}{prop}"

    return re.sub(
        r"(?P<indent>^[ \t]*)(?P<prop>(?:public|protected|private)\s+array\s+\$\w+)",
        repl,
        content,
        flags=re.MULTILINE,
    )


def patch_callables(content: str) -> str:
    signature_re = re.compile(
        r"(?P<doc>/\*\*(?:(?!\*/).)*\*/\s*)?"
        r"(?P<prefix>(?:public|protected|private|static)\s+(?:static\s+)?)?"
        r"function\s+(?P<name>\w+)\s*\((?P<params>[^)]*)\)"
        r"(?:\s*:\s*(?P<return>array))?",
        re.MULTILINE | re.DOTALL,
    )

    offset = 0
    parts: list[str] = []
    for match in signature_re.finditer(content):
        parts.append(content[offset : match.start()])
        doc = match.group("doc") or ""
        params_raw = match.group("params") or ""
        returns_array = match.group("return") == "array"

        array_params: list[str] = []
        for chunk in params_raw.split(","):
            chunk = chunk.strip()
            if not chunk:
                continue
            pm = re.search(r"array\s+\$(\w+)", chunk)
            if pm:
                array_params.append(pm.group(1))

        if not array_params and not returns_array:
            parts.append(match.group(0))
            offset = match.end()
            continue

        if doc:
            new_doc = inject_tags(doc, array_params, returns_array)
            parts.append(new_doc + match.group(0)[len(doc) :])
        else:
            indent_match = re.search(r"(?m)^([ \t]*)", content[: match.start()][::-1])
            indent = ""
            prefix = content[: match.start()]
            line_start = prefix.rfind("\n") + 1
            indent = prefix[line_start:]
            new_doc = f"/**"
            for param in array_params:
                new_doc += f"\n{indent} * @param {ARRAY_TYPE} ${param}"
            if returns_array:
                new_doc += f"\n{indent} * @return {ARRAY_TYPE}"
            new_doc += f"\n{indent} */"
            parts.append(f"{new_doc}\n{indent}{match.group(0)}")

        offset = match.end()

    parts.append(content[offset:])
    return "".join(parts)


def patch_file(path: Path) -> bool:
    original = path.read_text(encoding="utf-8")
    updated = patch_callables(patch_properties(original))
    if updated != original:
        path.write_text(updated, encoding="utf-8")
        return True
    return False


def main() -> int:
    changed = 0
    for base in TARGETS:
        for path in sorted(base.rglob("*.php")):
            if patch_file(path):
                changed += 1
                print(path.relative_to(ROOT))
    print(f"Patched {changed} files")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/fix-phpstan-iterables.php
<?php

declare(strict_types=1);

/**
 * Adds @param/@return/@var array<string, mixed> PHPDoc tags for bare array types
 * to satisfy PHPStan level 8 missingType.iterableValue checks.
 */

$root = dirname(__DIR__);
$paths = [
    $root . '/backend/app',
    $root . '/backend/bootstrap',
];

$arrayType = 'array<string, mixed>';

foreach ($paths as $basePath) {
    $iterator = new RecursiveIteratorIterator(
        new RecursiveDirectoryIterator($basePath, FilesystemIterator::SKIP_DOTS)
    );

    /** @var SplFileInfo $fileInfo */
    foreach ($iterator as $fileInfo) {
        if ($fileInfo->getExtension() !== 'php') {
            continue;
        }

        $path = $fileInfo->getPathname();
        $content = file_get_contents($path);
        if ($content === false) {
            continue;
        }

        $original = $content;
        $content = patchFile($content, $arrayType);

        if ($content !== $original) {
            file_put_contents($path, $content);
            echo "Patched: {$path}\n";
        }
    }
}

function patchFile(string $content, string $arrayType): string
{
    $content = patchPropertyDocblocks($content, $arrayType);
    $content = patchFunctionDocblocks($content, $arrayType);

    return $content;
}

function patchPropertyDocblocks(string $content, string $arrayType): string
{
    return (string) preg_replace_callback(
        '/((?:public|protected|private)\s+array\s+\$\w+)/m',
        static function (array $matches) use ($content, $arrayType): string {
            $propertyLine = $matches[1];
            $offset = strpos($content, $propertyLine);
            if ($offset === false) {
                return $propertyLine;
            }

            $before = substr($content, max(0, $offset - 600), $offset);
            if (preg_match('/@var\s+' . preg_quote($arrayType, '/') . '/s', $before) === 1) {
                return $propertyLine;
            }

            if (preg_match('/@var\s+array</s', $before) === 1) {
                return $propertyLine;
            }

            return "/** @var {$arrayType} */\n    " . $propertyLine;
        },
        $content
    );
}

function patchFunctionDocblocks(string $content, string $arrayType): string
{
    $pattern = '/(?P<signature>(?:public|protected|private|static)\s+(?:static\s+)?function\s+\w+\s*\((?P<params>[^)]*)\)(?:\s*:\s*(?P<return>array|void|string|int|bool|float|\?[\w\\\\|]+|[\w\\\\|]+))?)/m';

    return (string) preg_replace_callback(
        $pattern,
        static function (array $matches) use ($content, $arrayType): string {
            $signature = $matches['signature'];
            $params = $matches['params'] ?? '';
            $return = $matches['return'] ?? '';

            $offset = strpos($content, $signature);
            if ($offset === false) {
                return $signature;
            }

            $before = substr($content, max(0, $offset - 1200), $offset);
            $docStart = strrpos($before, '/**');
            $docEnd = $docStart !== false ? strrpos($before, '*/') : false;
            $hasDoc = $docStart !== false && $docEnd !== false && $docEnd > $docStart;

            $doc = $hasDoc ? substr($before, $docStart, $docEnd - $docStart + 2) : "/**\n */";

            $arrayParams = [];
            if ($params !== '') {
                foreach (explode(',', $params) as $param) {
                    $param = trim($param);
                    if ($param === '') {
                        continue;
                    }
                    if (preg_match('/array\s+\$(\w+)/', $param, $pm) === 1) {
                        $arrayParams[] = $pm[1];
                    }
                }
            }

            foreach ($arrayParams as $paramName) {
                if (preg_match('/@param\s+' . preg_quote($arrayType, '/') . '\s+\$' . preg_quote($paramName, '/') . '\b/', $doc) === 1) {
                    continue;
                }
                if (preg_match('/@param\s+array<[^>]+>\s+\$' . preg_quote($paramName, '/') . '\b/', $doc) === 1) {
                    continue;
                }
                $doc = preg_replace('/\*\/\s*$/', " * @param {$arrayType} \${$paramName}\n */", $doc) ?? $doc;
            }

            if ($return === 'array' && preg_match('/@return\s+' . preg_quote($arrayType, '/') . '\b/', $doc) !== 1
                && preg_match('/@return\s+array</', $doc) !== 1) {
                $doc = preg_replace('/\*\/\s*$/', " * @return {$arrayType}\n */", $doc) ?? $doc;
            }

            if (!$hasDoc && ($return === 'array' || $arrayParams !== [])) {
                return rtrim($doc) . "\n    " . $signature;
            }

            if ($hasDoc && $doc !== substr($before, $docStart, $docEnd - $docStart + 2)) {
                $oldDoc = substr($before, $docStart, $docEnd - $docStart + 2);
                return str_replace($oldDoc, $doc, $signature);
            }

            return $signature;
        },
        $content
    );
}
